src/evolve/downloader.go

```diff
@@ -0,0 +1,216 @@
+	self = new(Downloader)
+
+	self.Output = make(chan *model.ContractSource)
+
+	self.Task = task.NewTask(config, "evolve-downloader").
+		WithSubtaskFunc(self.run).
+		WithWorkerPool(config.Evolve.NumWorkers, config.Evolve.WorkerQueueSize).
+		WithOnAfterStop(func() {
+			close(self.Output)
+		})
+
+	return
+}
+
+func (self *Downloader) WithClient(client *arweave.Client) *Downloader {
+	self.client = client
+	return self
+}
+
+func (self *Downloader) WithInputChannel(v chan string) *Downloader {
+	self.input = v
+	return self
+}
+
+func (self *Downloader) run() (err error) {
+	
+	for srcId := range self.input {
+		srcId := srcId
+		var contractSrc *model.ContractSource
+		self.SubmitToWorker(func() {
+			tx, err := self.downloadContractSrcTx(self.client.GetTransactionById, srcId)
+			if err != nil {
+				self.Log.WithError(err).
+					WithField("srcId", srcId).
+					Error("Failed to download contract source transaction")
+				contractSrc.SrcTxId = srcId
+				contractSrc.Src.Set("error")
+				goto end
+			} 
+
+			contractSrc, err = self.loadContractSrcMetadata(tx)
+			if err != nil {
+				self.Log.WithError(err).
+					WithField("srcId", srcId).
+					Error("Failed to load contract source metadata")
+				contractSrc.SrcTxId = srcId
+				contractSrc.Src.Set("error")
+			}
+
+			end:
+				self.Output <- contractSrc
+		})
+	}
+
+	return nil
+}
+
+
+func (self *Downloader) downloadContractSrcTx(downloadTxFunc downloadTxFunc, srcId string) (out *arweave.Transaction, err error) {
+	err = task.NewRetry().
+		WithContext(self.Ctx).
+		WithMaxElapsedTime(0).
+		WithMaxInterval(self.Config.Evolve.SourceTransactionDownloadMaxInterval).
+		WithAcceptableDuration(self.Config.Evolve.SourceTransactionDownloadMaxInterval * 3).
+		WithOnError(func(err error, isDurationAcceptable bool) error {
+			// Permanent errors
+			if errors.Is(err, context.Canceled) && self.IsStopping.Load() {
+				return backoff.Permanent(err)
+			}
+
+			if errors.Is(err, arweave.ErrNotFound) {
+				return backoff.Permanent(err)
+			}
+
+			if errors.Is(err, arweave.ErrOverspend) {
+				return backoff.Permanent(err)
+			}
+
+			// Errors fo retry
+			self.Log.WithError(err).WithField("srcId", srcId).Warn("Failed to download contract source transaction, retrying after timeout")
+			
+			if errors.Is(err, arweave.ErrPending) {
+				time.Sleep(time.Second)
+				return err
+			}
+
+			if !isDurationAcceptable {
+				self.client.Reset()
+			}
+
+			return err
+		}).
+		Run(func() error {
+			out, err = downloadTxFunc(self.Ctx, srcId)
+			if err != nil {
+				self.Log.WithField("txId", srcId).Error("Failed to download contract source transaction")
+				return err
+			}
+			return err
+		})
+	if err != nil {
+		self.Log.WithError(err).WithField("txId", srcId).Error("Failed to download contract source transaction, giving up")
+		return
+	}
+	
+	return
+}
+
+func (self *Downloader) loadContractSrcMetadata(srcTx *arweave.Transaction) (out *model.ContractSource, err error) {
```

Comment:
If this code is the same as in Contract and it's not too much work make the code base common